File: Extra/GooglePlaces/ServiceManager.m
//
//  ServiceManager.m
//  venue
//
//  Created by Mitesh on 06/01/16.
//  Copyright © 2016 vue. All rights reserved.
//

#import "ServiceManager.h"

@implementation ServiceManager

#pragma mark - Private Initialization

- (id)init {
    self = [super init];
    
    if (self) {
        
    }
    
    return self;
}

#pragma mark - Default Manager

+ (ServiceManager *)sharedManager {
    static ServiceManager *_sharedManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedManager = [[self alloc] init];
    });
    
    return _sharedManager;
}


-(void)callWebServiceWithGET:(NSString *)webpath withTag:(NSString *)tagname params:(NSDictionary *)params{
    
    
    AFHTTPSessionManager *manager = [[AFHTTPSessionManager alloc]initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    
    [manager GET:webpath parameters:params progress:^(NSProgress * _Nonnull downloadProgress) {
        
    } success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        
        NSLog(@"Response : %@",[responseObject description]);
        NSLog(@"success!");
        
        [_delegate webServiceCallSuccess:responseObject forTag:tagname];

        
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        
        NSLog(@"error: %@", error);
        [_delegate webServiceCallFailure:error forTag:tagname];

        
    }];
    
    
}


-(void)callWebServiceWithPOST:(NSString *)webpath withTag:(NSString *)tagname params:(NSDictionary *)params{
    
    
    AFHTTPSessionManager *manager = [[AFHTTPSessionManager alloc]initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];

    [manager.responseSerializer setAcceptableContentTypes:[NSSet setWithObject:@"text/html"]];
    
    [manager POST:webpath parameters:params
         progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
             
             NSLog(@"Response : %@",[responseObject description]);
             NSLog(@"success!");
             
             [_delegate webServiceCallSuccess:responseObject forTag:tagname];
             
             
         } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
             
             NSLog(@"error: %@", error);
             
             [_delegate webServiceCallFailure:error forTag:tagname];
             
    }];
    
    
}

-(void)callWebServiceWithGETWithCustomHeader:(NSString *)webpath withTag:(NSString *)tagname params:(NSDictionary *)params withSecretKey:(NSString *)accessToken{
    
    
    AFHTTPSessionManager *manager = [[AFHTTPSessionManager alloc]initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    
    manager.requestSerializer = [AFHTTPRequestSerializer serializer];
    [manager.requestSerializer setValue:accessToken forHTTPHeaderField:@"X-AccessToken"];
    
    [manager GET:webpath parameters:params progress:^(NSProgress * _Nonnull downloadProgress) {
        
        
    } success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        
        NSLog(@"Response : %@",[responseObject description]);
        NSLog(@"success!");
        
        [_delegate webServiceCallSuccess:responseObject forTag:tagname];
        
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        
        NSLog(@"error: %@", error);
        
        [_delegate webServiceCallFailure:error forTag:tagname];

        
    }];
    
}

-(void)callWebServiceWithPOSTWithCustomHeader:(NSString *)webpath withTag:(NSString *)tagname params:(NSDictionary *)params withSecretKey:(NSString *)accessToken{
    
    
    AFHTTPSessionManager *manager = [[AFHTTPSessionManager alloc]initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    
    manager.requestSerializer = [AFHTTPRequestSerializer serializer];
    [manager.requestSerializer setValue:accessToken forHTTPHeaderField:@"X-AccessToken"];
    
    [manager POST:webpath parameters:params
         progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
             
             NSLog(@"Response : %@",[responseObject description]);
             NSLog(@"success!");
             
             [_delegate webServiceCallSuccess:responseObject forTag:tagname];
             
             
         } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
             
             NSLog(@"error: %@", error);
             
             [_delegate webServiceCallFailure:error forTag:tagname];
             
         }];
    
    
}

-(void)callWebServiceWithPOST:(NSString *)webpath withTag:(NSString *)tagname params:(NSArray *)params imgArray:(NSMutableArray *)arrImages{
    
    NSMutableDictionary *parameters = [[NSMutableDictionary alloc] init];
    for(NSDictionary *currentDic in params){
        [parameters addEntriesFromDictionary:currentDic];
    }
    
    NSLog(@"\n%@", webpath);
    NSLog(@"%@\n", parameters);
    
    NSMutableURLRequest *request  = [[AFHTTPRequestSerializer serializer] multipartFormRequestWithMethod:@"POST" URLString:webpath parameters:parameters constructingBodyWithBlock:^(id<AFMultipartFormData> formData) {
        
       
        
        for(int i = 0; i < [arrImages count]; i++){
            
            /* AttachmentViewModel *attachmentModel = arrImages[i];
             UIImage *img = attachmentModel.Image;
             NSData *imgData = [self CompressImage:img];
             //NSString *name = attachmentModel.ImageFileName;
             NSString *name = @"Profile_pic";*/
            
            UIImage *img = arrImages[i];
            NSData *imgData = UIImageJPEGRepresentation(img, 0.5);
            NSInteger imageSize   = imgData.length;
            NSLog(@"SIZE OF IMAGE: %.2f Mb", (float)imageSize/1024/1024);
            
            
            UIImage *compressedImage = [UIImage imageWithData:imgData];
            [formData appendPartWithFileData:UIImagePNGRepresentation(compressedImage) name:@"file" fileName:@"profile_pic.png" mimeType:@"image/png"];
            
            
            // [formData appendPartWithFileData:imgData name:name/*@"uploadedfile"*/ fileName:name mimeType:@"image/png"];
            
        }
        
        NSLog(@"%@", formData);
        
    } error:nil];
    
 
    
    AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    manager.responseSerializer = [AFHTTPResponseSerializer serializer];

    [request setValue:@"va2f40cc914983764463b4477ac33e" forHTTPHeaderField:@"Apz-AppId"];
    [request setValue:@"Basic dmFAZWJpenppbmZvdGVjaC5jb206MTIzNDU2Nzg=" forHTTPHeaderField:@"Apz-Token"];
    [request setValue:@"application/json" forHTTPHeaderField:@"accept"];
    
    NSURLSessionUploadTask *uploadTask = [manager
                                          uploadTaskWithStreamedRequest:request
                                          progress:^(NSProgress * _Nonnull uploadProgress) {
                                              
                                              dispatch_async(dispatch_get_main_queue(), ^{
                                                  //Update the progress view
                                                  NSLog(@"Completed : %f", uploadProgress.fractionCompleted);
                                              });
                                          }
                                          completionHandler:^(NSURLResponse * _Nonnull response, id  _Nullable responseObject, NSError * _Nullable error) {
                                              if (error) {
                                                  if ([self.delegate respondsToSelector:@selector(webServiceCallFailure:forTag:)]) {
                                                      [self.delegate webServiceCallFailure:error forTag:tagname];
                                                  }
                                              } else {
                                                  NSError *err;
                                                  id responseFromServer = [NSJSONSerialization JSONObjectWithData:responseObject options:NSJSONReadingMutableContainers error:&err];
                                                  if(err){
                                                      NSLog(@"Error : %@", err.localizedDescription);
                                                      NSString *strResponse = [[NSString alloc] initWithData:responseObject encoding:NSUTF8StringEncoding];
                                                      NSLog(@"Response : %@", strResponse);
                                                  }
                                                  [self.delegate webServiceCallSuccess:responseFromServer forTag:tagname];
                                              }
                                              
                                          }];
    
    [uploadTask resume];
}



@end

File: Objective C/Extra/CommonViewModel/Common/ServiceManager.swift
//
//  ServiceManager.swift
//  SwiftProjectStructure
//
//  Created by Krishna Patel on 20/06/18.
//  Copyright © 2018 Krishna. All rights reserved.
//

import UIKit
import Alamofire
import SwiftyJSON
import AlamofireSwiftyJSON

@objc protocol ServiceManagerDelegate: NSObjectProtocol{
    func webServiceCallSuccess(_ response: Any?, forTag tagname: String?)
    func webServiceCallFailure(_ error: Error?, forTag tagname: String?)
}

class ServiceManager: NSObject {

    var delegate:ServiceManagerDelegate?

    //GET Service Calling
    func callWebServiceWithGET(webpath: String?, withTag tagname: String?) {
        
        Alamofire.request(webpath!, method: .get, parameters: nil, encoding: URLEncoding.httpBody, headers: nil).authenticate(user: "", password: "").responseSwiftyJSON(completionHandler:{ (response) in
            
           //print(response.request)
           //print(response.result.value)
            
            if let json = response.result.value {
                guard !json["data"].isEmpty else{
                    return
                }
                guard let rawData = try? json.rawData() else {
                    return
                }

                var dicResponse : [String : AnyObject] = [:]
                do {
                    dicResponse =  (try JSONSerialization.jsonObject(with: rawData, options: []) as? [String:AnyObject])!
                    print(dicResponse)
                } catch let error as NSError {
                    print(error)
                 }
             
                if (self.delegate?.responds(to: #selector(self.delegate?.webServiceCallSuccess(_:forTag:))))! {
                    self.delegate?.webServiceCallSuccess(dicResponse, forTag: tagname)
                }
            }
            else{
                if (self.delegate?.responds(to: #selector(self.delegate?.webServiceCallFailure(_:forTag:))))! {
                    self.delegate?.webServiceCallFailure(response.result.error, forTag: tagname)
                }

            }
            
        })
    }
    
    //POST Service Calling
    func callWebServiceWithPOST(webpath: String?, withTag tagname: String?, params: Parameters) {
        
        Alamofire.request(webpath!, method: .post, parameters: params, encoding: URLEncoding.httpBody, headers: nil).authenticate(user: "", password: "").responseSwiftyJSON(completionHandler:{ (response) in
            
            //print(response.request)
            //print(response.result)
            
            if let json = response.result.value { 
              /*  guard !json["data"].isEmpty else{
                    return
                }*/
                guard let rawData = try? json.rawData() else {
                    return
                }
                
                var dicResponse : [String : AnyObject] = [:]
                do {
                    dicResponse =  (try JSONSerialization.jsonObject(with: rawData, options: []) as? [String:AnyObject])!
                    print(dicResponse)
                } catch let error as NSError {
                    print(error)
                }
                
                if (self.delegate?.responds(to: #selector(self.delegate?.webServiceCallSuccess(_:forTag:))))! {
                    self.delegate?.webServiceCallSuccess(dicResponse, forTag: tagname)
                }
                
            }
            else{
                if (self.delegate?.responds(to: #selector(self.delegate?.webServiceCallFailure(_:forTag:))))! {
                    self.delegate?.webServiceCallFailure(response.result.error, forTag: tagname)
                }
            }
            
        })
    }
    
    func callWebServiceWithPOST(webpath: String?, withTag tagname: String?, params: Parameters, imgArray:[AttachmentViewModel])  {
        let unit64:UInt64 = 10_000_000
        Alamofire.upload(multipartFormData: { (multipartFormData) in
           
            for model in imgArray  {
                let imgData = UIImageJPEGRepresentation(model.Image, 0.7)
                multipartFormData.append(imgData!, withName: model.ImageFileName, fileName: model.ImageFileName, mimeType: "image/png")
            }
            
            for (key, value) in params {
                multipartFormData.append((value as AnyObject).data(using:String.Encoding(rawValue: String.Encoding.utf8.rawValue).rawValue)!, withName: key)
            }
            
        }, usingThreshold: unit64, to: webpath!, method: .post, headers: nil, encodingCompletion: { (encodingResult) in
            print("encoding result:\(encodingResult)")
            switch encodingResult {
            case .success(let upload, _, _):
                upload.uploadProgress(closure: { (Progress) in
                    print("Upload Progress: \(Progress.fractionCompleted)")
                    //send progress using delegate
                })
                upload.responseSwiftyJSON(completionHandler: { (response) in
                    print("response:==>\(response)")
                    
                    if let json = response.result.value {
//                        guard !json["data"].isEmpty else{
//                            return
//                        }
                        guard let rawData = try? json.rawData() else {
                            return
                        }
                        
                        var dicResponse : [String : AnyObject] = [:]
                        do {
                            dicResponse =  (try JSONSerialization.jsonObject(with: rawData, options: []) as? [String:AnyObject])!
                            print(dicResponse)
                        } catch let error as NSError {
                            print(error)
                        }
                        
                        if (self.delegate?.responds(to: #selector(self.delegate?.webServiceCallSuccess(_:forTag:))))! {
                            self.delegate?.webServiceCallSuccess(dicResponse, forTag: tagname)
                        }
                        
                    }
                })
                
            case .failure(let encodingError):
                if (self.delegate?.responds(to: #selector(self.delegate?.webServiceCallFailure(_:forTag:))))! {
                    self.delegate?.webServiceCallFailure(encodingError, forTag: tagname)
                }
            }
        })
    }
}
